src/coreclr/vm/comsynchronizable.cpp

```diff
@@ -703,23 +619,22 @@ void ReleaseThreadExternalCount(Thread * pThread)
 typedef Holder<Thread *, DoNothing, ReleaseThreadExternalCount> ThreadExternalCountHolder;
 
 // Wait for the thread to die
-BOOL ThreadNative::DoJoin(THREADBASEREF DyingThread, INT32 timeout)
+static BOOL DoJoin(QCall::ThreadHandle dyingThreadHandle, INT32 timeout)
 {
     CONTRACTL
     {
         THROWS;
         GC_TRIGGERS;
-        MODE_COOPERATIVE;
-        PRECONDITION(DyingThread != NULL);
+        MODE_PREEMPTIVE;
         PRECONDITION((timeout >= 0) || (timeout == INFINITE_TIMEOUT));
     }
     CONTRACTL_END;
 
-    Thread * DyingInternal = DyingThread->GetInternal();
+    Thread * DyingInternal = dyingThreadHandle;
 
     // Validate the handle.  It's valid to Join a thread that's not running -- so
     // long as it was once started.
-    if (DyingInternal == 0 ||
+    if (DyingInternal == NULL ||
```

Comment:
This cannot be NULL now. `GetNativeHandle` on the managed side will throw a different exception for null thread handle.

Reply:
Reverted the managed side changes.